### PR TITLE
refactor: deploy all libraries when running tests

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Closes https://github.com/foundry-rs/foundry/issues/1161 
Closes https://github.com/foundry-rs/foundry/issues/6120
Closes https://github.com/foundry-rs/foundry/issues/8014

Keeping #4049 open as this still won't work in scripts.

## Solution

Instead of linking each contract separately and keeping separate `libs_to_deploy` for each test contract we now link entire project output and keep single `libs_to_deploy` and `known_contracts`. That way, we reduce memory usage and by deploying all needed libraries unblock usage of `getCode`/`getDeployedCode` with artifacts requiring linking.

To avoid nonce of the `sender` being affected by the number of external libraries used in project, I've inroduced separate `LIBRARY_DEPLOYER` which deploys all libraries, that way, test contract will always have the same address regardless of the number of libraries it requires to run.

### CODE REVIEW

File: crates/forge/bin/cmd/test/mod.rs [hunk 1]
```diff
@@ -407,7 +408,7 @@ impl TestArgs {
 
             // Build the trace decoder.
             let mut builder = CallTraceDecoderBuilder::new()
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:410]: In this case can we hoist these builders out of the loop again?

File: crates/forge/src/runner.rs [hunk 2]
```diff
@@ -43,12 +43,21 @@ use std::{
     time::Instant,
 };
 
+/// When running tests, we deploy all external libraries present in the project. To avoid additional
+/// libraries affecting nonces of senders used in tests, we are using separsate address to
```
Comment (DaniPopes) [inline crates/forge/src/runner.rs:47]:
```suggestion
/// libraries affecting nonces of senders used in tests, we are using separate address to
```

File: crates/forge/src/runner.rs [hunk 3]
```diff
@@ -43,12 +43,21 @@ use std::{
...
+/// predeploy libraries.
+///
+/// address(uint160(uint256(keccak256("foundry library deployer"))))
```
Comment (DaniPopes) [inline crates/forge/src/runner.rs:50]:
```suggestion
/// `address(uint160(uint256(keccak256("foundry library deployer"))))`
```

**Comment by DaniPopes** [comment]
This should close #8014

File: crates/forge/bin/cmd/test/mod.rs [hunk 4]
```diff
@@ -392,31 +416,10 @@ impl TestArgs {
...
-                .with_verbosity(verbosity);
-            // Signatures are of no value for gas reports.
-            if !self.gas_report {
-                builder = builder.with_signature_identifier(SignaturesIdentifier::new(
-                    Config::foundry_cache_dir(),
-                    config.offline,
-                )?);
-            }
-            let mut decoder = builder.build();
+            let mut decoder = decoder.clone();
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:422]: Do we need this clone? I added `decoder.clear_addresses();` so we can reuse the same decoder across tests

File: crates/forge/bin/cmd/test/mod.rs [hunk 5]
```diff
@@ -376,6 +377,29 @@ impl TestArgs {
         }
 
         let remote_chain_id = runner.evm_opts.get_remote_chain_id().await;
+        let known_contracts = runner.known_contracts.clone();
+
+        // Set up trace identifiers.
+        let mut identifier = TraceIdentifiers::new().with_local(&known_contracts);
```
Comment (DaniPopes) [inline crates/forge/bin/cmd/test/mod.rs:383]: I'd move these after spawning the runner thread for faster start up times

File: crates/forge/bin/cmd/test/mod.rs (hunk 4 above)
Comment (klkvr) [reply crates/forge/bin/cmd/test/mod.rs]: ah right, it seemed that we needed ownership on it for `suite_result.last_run_decoder`

overlooked that we can just set it after the loop

File: crates/forge/bin/cmd/test/mod.rs (hunk 5 above)
Comment (klkvr) [reply crates/forge/bin/cmd/test/mod.rs:391]: yep, makes sense

addressed both comments in a88df74

**Review by DaniPopes — APPROVED** [review]
nice

**Review by mattsse — APPROVED** [review]
awesome

**Pull request merged**